### DISCUSSION TITLE
Handling different urls for internal vs external access to Keycloak.

### DISCUSSION THREAD

**Discussion by dusktreader** [keycloak/keycloak#35302] [Q&A]
My current project is still very much in development, and not deployed anywhere other than my local machines using docker compose. However, I think it might be an issue that could be faced in "real" environments where access to Keycloak is affected by DNS.

In my situation, I have a set of services deployed in docker compose along with a keycloak instance and a traefik reverse proxy. The APIs are accessed via requests to the host machine (my desktop) and routed via traefik to their docker containers. Additionally, I have a route established for keycloak as well. I can successfully connect with Keycloak (from a command line) interface through it's traefik-configured url and obtain tokens.

However, the containers hosting my APIs connect to Keycloak via the docker network. They use the docker dns to resolve the hostname for keycloak. This does not match the "frontend url" that keycloak is configured with, and thus they can't retrieve JWKs from Keycloak in order to authenticate incoming requests. This happens because the "internal" API can successfully read the OpenID configuration from keycloak using the hostname known in the docker network ("keycloak"). However, the config that is fetched includes URLs where the Keycloak "frontend url" is provided ("raven/shrump/keycloak"), and the container can't resolve hostname "raven".

I am trying to figure out the best way to configure all of this. Ideally, it would not require me to add a hostfile entry for my API containers nor some mechanism to make the containers aware of their host machine's IP/hostname.

Is there a way to configure a different "frontend url" for internal access (from within the docker network) versus external access (routed through traefik reverse proxy)?

For reference:

From `docker-compose.yaml`
```file=docker-compose.yaml
# Trimmed for relevance and brevity
services:
  pin-reader-api:
    environment:
      - ARMASEC_DOMAIN=keycloak:8080/realms/master
      - ARMASEC_AUDIENCE=https://shrump.dusktreader.dev
      - ARMASEC_DEBUG=true
      - ARMASEC_USE_HTTPS=false
    depends_on:
      keycloak:
        condition: service_healthy
    networks:
      - internal-net
      - wicket-net

  keycloak:
    image: keycloak/keycloak:26.0
    restart: always
    volumes:
      - kc-realm-files:/opt/keycloak/data/import/
      - ./etc/keycloak-local.json:/opt/keycloak/data/import/keycloak-local.json
    environment:
      - KC_BOOTSTRAP_ADMIN_USERNAME=admin
      - KC_BOOTSTRAP_ADMIN_PASSWORD=admin
      - KC_HEALTH_ENABLED=true
      - KC_DB=postgres
      - KC_DB_URL=jdbc:postgresql://keycloak-db:5432/keycloak
      - KC_DB_USERNAME=keycloak
      - KC_DB_PASSWORD=keycloak
      - KC_HOSTNAME=http://raven/shrump/keycloak
      - KC_HOSTNAME_STRICT=false
      - KC_HOSTNAME_STRICT_HTTPS=false
      - KC_PROXY_HEADERS=xforwarded
    command:
      - "start-dev"
      - "--import-realm"
    depends_on:
      keycloak-db:
        condition: service_healthy
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=wicket-net"
      - "traefik.http.middlewares.shrump-keycloak.stripprefix.prefixes=/shrump/keycloak"
      - "traefik.http.routers.shrump-keycloak.middlewares=shrump-keycloak"
      - "traefik.http.routers.shrump-keycloak.rule=PathPrefix(`/shrump/keycloak`)"
      - "traefik.http.routers.shrump-keycloak.entrypoints=web"
      - "traefik.http.services.shrump-keycloak.loadbalancer.server.port=8080"
    networks:
      - wicket-net
      - internal-net
   
  traefik:
    image: "traefik:v3.2"
    container_name: "traefik"
    command:
      - "--configfile=/etc/traefik/traefik.yaml"
    ports:
      - "80:80"
      - "8080:8080"
    volumes:
      - "./etc/:/etc/traefik/"
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
    networks:
      - wicket-net

networks:
  wicket-net:
    external: true
  internal-net:
```

```file=traefik.yaml
api:
  insecure: true
  dashboard: true

providers:
  docker:
    watch: true
    exposedByDefault: false

log:
  level: DEBUG
  format: common

entryPoints:
  web:
    address: ":80"
```

From `logs.txt`
```file=logs.txt
pin-reader-api-1  | 2024-11-23 23:19:52.019 | DEBUG    | armasec.token_security:_load_manager:220 - Lazy loading TokenManager for domain keycloak:8080/realms/master
pin-reader-api-1  | 2024-11-23 23:19:52.020 | DEBUG    | armasec.openid_config_loader:jwks:97 - Fetching jwks
pin-reader-api-1  | 2024-11-23 23:19:52.020 | DEBUG    | armasec.openid_config_loader:config:78 - Fetching openid configration
pin-reader-api-1  | 2024-11-23 23:19:52.020 | DEBUG    | armasec.openid_config_loader:_load_openid_resource:59 - Attempting to fetch from openid resource 'http://keycloak:8080/realms/master/.well-known/openid-configuration'
pin-reader-api-1  | 2024-11-23 23:19:52.078 | DEBUG    | armasec.openid_config_loader:_load_openid_resource:59 - Attempting to fetch from openid resource 'http://raven/shrump/keycloak/realms/master/protocol/openid-connect/certs'
pin-reader-api-1  | 2024-11-23 23:19:52.108 | DEBUG    | armasec.utilities:log_error:45 - Call to url http://raven/shrump/keycloak/realms/master/protocol/openid-connect/certs failed -- ConnectError: [Errno -5] No address associated with hostname
```

You can see here that the API can successfully read the OpenID configuration from keycloak using the hostname known in the docker network ("keycloak"). However, the config that is fetched includes URLs where the Keycloak "frontend url" is provided ("raven/shrump/keycloak"), and the container can't resolve hostname "raven".

Any advice here greatly appreciated!

**Comment by embesozzi** [comment]
See [34712](https://github.com/keycloak/keycloak/discussions/34712)
